### PR TITLE
Add video classification fine-tuning notebook

### LINKED ISSUE #258: Add a notebook for fine-tuning video classification
We have a couple of video models: https://huggingface.co/models?pipeline_tag=video-classification&sort=downloads. 

Considering videos are an essential modality to tackle with lots of exciting applications, it is a good idea to have a notebook tutorial showing how to fine-tune a video classification model. 

We don't have involved support for videos in `datasets` yet ([issue](https://github.com/huggingface/datasets/issues/5032)), so, the tutorial will rely on `decord` and `torchvision` and will also leverage different utilities from `datasets` when necessary. I also understand that videos in `transformers` are new and may go through different transition phases. So, when there's better support for videos in `datasets` I can update the notebook as well.

### PR DESCRIPTION
**Pull request by sayakpaul**
Closes #258.

Colab Notebook: https://colab.research.google.com/drive/1xMpzrFPH7c4hOEb92zz2EneQfvxZVinn?usp=sharing

Spaces: https://huggingface.co/spaces/sayakpaul/video-classification-ucf101-subset

Inline demo:

https://user-images.githubusercontent.com/22957388/200532344-01dcb006-9054-4a7e-8c3f-febcb0bea3ee.mov

After the PR is ready to be merged, I will add a PR to include the notebook GIF on Hub.

### CODE REVIEW

**Comment by nateraw** [comment]
Watching this. please ping when its ready

**Comment by sayakpaul** [comment]
@NielsRogge @nateraw up for reviews.

**Comment by nateraw** [comment]
<p>+1</p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/261/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nateraw** [comment]
<p>Yea I've never seen this before. Something wonky going on...haven't really looked but I'm guessing its either:</p><p> </p><ul><li>providing <code>max_steps</code> but trying to use epoch strategy for save and eval, or</li><li>something to do with the iterable dataset </li></ul>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/261/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by nateraw** [comment]
Seems the large epoch number is coming from here:
https://github.com/huggingface/transformers/blob/4eb918e656944df2757513c535e8ad8c01d632e2/src/transformers/trainer.py#L1556

**Comment by sayakpaul** [comment]
<p>Done.</p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/261/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>No, I didn't use the eqn to calculate it. Just went with the defaults of PyTorch Video examples. </p><p> </p><p>If go with <code>floor(clip_duration = num_frames * sample_rate / fps)</code> then it's 2.   </p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/261/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
<p>👍</p>

---

 <a href='https://app.reviewnb.com/huggingface/notebooks/pull/261/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by sayakpaul** [comment]
> Oh, also any reason why you're not using the base model, but instead the kinetics fine tuned model as the starting point?

Good point. Let me actually try it out.

**Comment by sayakpaul** [comment]
> https://github.com/huggingface/transformers/blob/4eb918e656944df2757513c535e8ad8c01d632e2/src/transformers/trainer.py#L1556

Anything you'd suggest to modify in the current code? IMO, we could just add a note on it in the notebook for now:

> Don't worry about the astronomical number of epochs. Since we have specified `max_steps`, the model will be trained for so. 

WDYT?

**Comment by sayakpaul** [comment]
@nateraw incorporated [this](https://github.com/huggingface/notebooks/pull/261#issuecomment-1308064071) in https://github.com/huggingface/notebooks/pull/261/commits/d504f9b00ea3c1651e698d4a1a6889b5c1dd28f2. Let me know what you think.

**Comment by nateraw** [comment]
Ah I see @sayakpaul - we are seeing lower accuracy with that, right? kinetics helps a lot because domain is similar. Maybe switch it back then, I guess? idk, not really too picky.

Also, technically the eval method is a tad off, as we should be doing some X number of clips with X different crops (described in paper). but I really don't think that's something we should worry about here.

**Comment by sayakpaul** [comment]
> Ah I see @sayakpaul - we are seeing lower accuracy with that, right? kinetics helps a lot because domain is similar. Maybe switch it back then, I guess? idk, not really too picky.
> 
> Also, technically the eval method is a tad off, as we should be doing some X number of clips with X different crops (described in paper). but I really don't think that's something we should worry about here.

@nateraw excellent points! Thank you! I have added more elaborate notes in the notebook to reflect these. IMO it's okay to leave some homework for the readers to encourage them to experiment with different settings. LMKWYT.

**Comment by nateraw** [comment]
Sounds good! Will give this a final look over today

**Review by nateraw — CHANGES_REQUESTED** [review]
Please update to use 2.13 for `clip_duration` if you aren't going to update to use the snippet I sent over. this is fine and is what I've seen other notebooks do.

---

Found a bigger issue though...

Your dummy dataset has leakage of scenes across the splits. Essentially we are training and testing on examples from the same scene video, so the model is seeing test data at train time.

The file names have a scene/group ID (`g02` below) and a clip ID (`c01` below)

`ApplyLipstick/v_ApplyLipstick_g02_c01.avi`

You want to make sure scenes used in one split do not end up in another split.

Screen Shot 2022-11-09 at 8 35 40 PM

You can use one of the official annotations splits ([download link](https://www.crcv.ucf.edu/data/UCF101/UCF101TrainTestSplits-RecognitionTask.zip)) if you don't want to do this yourself. this has only train and test though.

**Comment by sayakpaul** [comment]
Thanks for spotting this @nateraw! I remember Jeremy Howard taking the exact same example to explain data leakage in one of the cohorts of the seminal course. 

I will work on it to update things based on your suggestions.

**Review by nateraw — APPROVED** [review]
LGTM! Thanks for your hard work on this and iterating on review.

**Comment by sayakpaul** [comment]
Thanks for your rigorous review, Nate!

**Comment by bayesianist** [comment]
Hi Paul, there is an error about upsample_bilinear2d() received an invalid combination of arguments on the train_results = trainer.train() step. Could you have a look ? It seems like it does not work on 1.11 and nightly anymore

**Comment by bayesianist** [comment]
Finally, I think there is a missing code to define resize_to variable
if "shortest_edge" in image_processor.size:
height = width = image_processor.size["shortest_edge"]
else:
height = image_processor.size["height"]
width = image_processor.size["width"]
resize_to = (height, width)
